core/src/main/scala/com/pingcap/tispark/auth/TiAuthorization.scala

```diff
@@ -0,0 +1,282 @@
+  }
+
+  {
+    TiAuthorization.dbPrefix = tiConf.getDBPrefix
+    val option = new TiDBOptions(parameters)
+    try {
+      this.tiDBJDBCClient = new TiDBJDBCClient(TiDBUtils.createConnectionFactory(option.url)())
+    } catch {
+      case e: Throwable => {
+        logger.error(f"Failed to create tidb jdbc client with url ${option.url}", e)
+        System.exit(-1)
+      }
+    }
+
+    task.run()
+    scheduler.scheduleWithFixedDelay(task, refreshInterval, refreshInterval, TimeUnit.SECONDS)
+  }
+
+  def getPrivileges: PrivilegeObject = {
+    var input = JavaConverters.asScalaBuffer(tiDBJDBCClient.showGrants).toList
+
+    /** TODO: role-based privilege
+     *
+      * Show grants using more than two roles return incorrect result
+     * https://github.com/pingcap/tidb/issues/30855
+     *
+      * val roles = extractRoles(input)
+     * input =
+     * if (roles.nonEmpty)
+     * JavaConverters
+     * .asScalaBuffer(
+     * tiDBJDBCClient.showGrantsUsingRole(roles.asJava)
+     * ).toList
+     * else input
+     */
+
+    parsePrivilegeFromRow(input)
+  }
+
+  def getPDAddress(): String = {
+    try {
+      tiDBJDBCClient.getPDAddress
+    } catch {
+      case e: Throwable =>
+        throw new IllegalArgumentException(
+          "Failed to get pdAddress from TiDB, please make sure user has `PROCESS` privilege on `INFORMATION_SCHEMA`.`CLUSTER_INFO`",
+          e)
+    }
+  }
+
+  def checkGlobalPiv(mySQLPriv: MySQLPriv.Value): Boolean = {
+    val privs = globalPriv.get()
+    if (privs.contains(MySQLPriv.AllPriv) || privs.contains(mySQLPriv)) true
+    else false
+  }
+
+  def checkDatabasePiv(db: String, mySQLPriv: MySQLPriv.Value): Boolean = {
+    val privs = databasePrivs.get().getOrElse(db, List())
+    if (privs.contains(MySQLPriv.AllPriv) || privs.contains(mySQLPriv)) true
+    else false
+  }
+
+  def checkTablePiv(db: String, table: String, mySQLPriv: MySQLPriv.Value): Boolean = {
+    val privs =
+      tablePrivs.get().getOrElse(db.trim, Map()).getOrElse(table.trim, List())
+    if (privs.contains(MySQLPriv.AllPriv) || privs.contains(mySQLPriv)) true
+    else false
+  }
+
+  def checkPrivs(db: String, table: String, requiredPriv: MySQLPriv.Value): Unit = {
+    if (!checkGlobalPiv(requiredPriv) && !checkDatabasePiv(db, requiredPriv) && !checkTablePiv(
+        db,
+        table,
+        requiredPriv))
+      throw new SQLException(f"Lack of privilege:$requiredPriv on database:$db table:$table")
+  }
+
+  def visible(db: String, table: String): Boolean = {
+    // Account who has ShowDBPriv is able to see all databases and tables regardless of revokes.
+    if (globalPriv.get().contains(MySQLPriv.AllPriv) || globalPriv
+        .get()
+        .contains(MySQLPriv.ShowDBPriv)) {
+      return true
+    }
+
+    // Account who has any Priv of the database/table can see the database/table
+    if (table.isEmpty) {
+      databasePrivs.get().keySet.contains(db) || tablePrivs
+        .get()
```

Comment:
I'm not familiar with scala Map. Can it do something like `databasePrivs.get().contains(db)`?

Reply:
done